content/2022-04-29-firmware-updates-part-1/index.md

```diff
@@ -0,0 +1,160 @@
+
+Once the new firmware is written, and marked to be swapped with the current firmware, the application needs to reset the device. Upon reset, the device will enter the bootloader once again, and attempt to swap the firmware before booting the next version.
+
+NOTE: By structuring your application accordingly, you may update the firmware in parallel while running other tasks. This reduces the downtime of your application to a minimum.
+
+## Swap algorithm
+
+At the core of the bootloader is the bank swapping algorithm, which is not tied to any specific platform. The algorithm keeps an internal state of the copy progress in flash, using 1 word per page to spread the writes.
+
+Lets assume a flash size of 3 pages for the Active partition, and 4 pages for the DFU partition. The swap index contains the copy progress, as to allow detecting if copy is completed or not on power failure. The index counter is represented within 1 or more pages (depending on total flash size), where a page X is considered swapped if index at location (X + WRITE_SIZE) contains a non-erased value. This ensures that index updates can be performed atomically and avoid a situation where the wrong index value is set (page write size is "atomic").
+
+The initial state of the Active, DFU and State partitions are shown below:
+
+<figure>
+    <img src="swap_initial.png" alt="Initial state" />
+    <figcaption>Initial State</figcaption>
+</figure>
+
+The `State` partition is only showing the index for the `Active` partition for simplicity. The algorithm starts by copying 'backwards', and after the first step, the layout is as follows:
+
+<figure>
+    <img src="swap_step1.png" alt="Copy state 1" />
+    <figcaption>Copy State 1</figcaption>
+</figure>
+
+The next iteration performs the same steps:
+
+<figure>
+    <img src="swap_step2.png" alt="Copy state 2" />
+    <figcaption>Copy State 2</figcaption>
+</figure>
+
+And again until we're done:
+
+<figure>
+    <img src="swap_final.png" alt="Final state" />
+    <figcaption>Final State</figcaption>
+</figure>
+
+The reverting algorithm uses the swap index to check if images were swapped, or that the application failed to mark the boot successful. In this case, the revert algorithm will
+run.
+
+The revert index is located separately from the swap index, to ensure that revert can continue on power failure.
+
+Another thing to note is that the revert algorithm works forwards, by starting copying into the 'unused' DFU page at the start.
+
+Once the swap process is complete, the bootloader may jump to the application at beginning of the active partition.
+
+This is a platform-specific step done in the `embassy-boot-stm32` or `embassy-boot-nrf` part of the bootloader.
+
+NOTE: The new application is responsible for marking itself as successfully booted, otherwise the bootloader will attempt to revert to the previous application when restarted!
```

Comment:
Or maybe *here* for the "mark as successful" example.

Everyone loves code examples!